internal/vulnstore/postgres/getupdateoperations.go

```diff
@@ -0,0 +1,150 @@
+package postgres
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/rs/zerolog"
+
+	"github.com/quay/claircore/libvuln/driver"
+)
+
+// GetLatestUpdateRef implements driver.Updater.
+func (s *Store) GetLatestUpdateRef(ctx context.Context) (uuid.UUID, error) {
+	const query = `SELECT ref FROM update_operation ORDER BY id USING > LIMIT 1;`
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "internal/vulnstore/postgres/getLatestRef").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	var ref uuid.UUID
+	if err := s.pool.QueryRow(ctx, query).Scan(&ref); err != nil {
+		return uuid.Nil, err
+	}
+	return ref, nil
+}
+
+func getLatestRefs(ctx context.Context, pool *pgxpool.Pool) (map[string]uuid.UUID, error) {
+	const query = `SELECT updater, ref FROM update_operation GROUP BY updater ORDER BY updater, id USING > LIMIT 1;`
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "internal/vulnstore/postgres/getLatestRefs").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	rows, err := pool.Query(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make(map[string]uuid.UUID)
+	var u string
+	var id uuid.UUID
+	for rows.Next() {
+		if err := rows.Scan(&u, &id); err != nil {
+			return nil, err
+		}
+		ret[u] = id
+	}
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+	log.Debug().
+		Int("count", len(ret)).
+		Msg("found updaters")
+	return ret, nil
+}
+
+func getUpdateOperations(ctx context.Context, pool *pgxpool.Pool, updater ...string) (map[string][]driver.UpdateOperation, error) {
+	const (
+		query       = `SELECT ref, updater, fingerprint, date FROM update_operation WHERE updater = $1 ORDER BY id DESC;`
+		getUpdaters = `SELECT DISTINCT(updater) FROM update_operation;`
+	)
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "internal/vulnstore/postgres/getUpdateOperations").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	tx, err := pool.Begin(ctx)
+	defer tx.Rollback(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to being transaction: %w", err)
+	}
+	out := make(map[string][]driver.UpdateOperation)
+
+	// Get distinct updaters from database if nothing specified.
+	if len(updater) == 0 {
+		updater = []string{}
+		rows, err := tx.Query(ctx, getUpdaters)
+		switch {
+		case err == nil:
+		case errors.Is(err, pgx.ErrNoRows):
+			rows.Close()
+		default:
+			rows.Close()
+			return nil, fmt.Errorf("failed to get distinct updates: %w", err)
+		}
+		// use a closure to defer rows.Close() and ensure connection can be re-used
+		// see: https://pkg.go.dev/github.com/jackc/pgx?tab=doc#Rows
+		err = func() error {
```

Comment:
I saw you use another pattern above, does defering rows.Close and then explicitly calling rows.Close() if an early return is not necessary apply here?

Reply:
This is code just carried forward. I can update.